src/peer_channels.rs

```diff
@@ -0,0 +1,597 @@
+                },
+                |g| {
+                    g.map(|data| {
+                        let (key, server) = data?;
+                        let server = server.map(|data| Ok::<_, url::ParseError>(RelayUrl::from(Url::parse(&data)?))).transpose()?;
+                        let id = NodeId::from_bytes(&key.try_into()
+                        .map_err(|_| anyhow!("Can't convert sql data to [u8; 32]"))?)?;
+                        Ok::<_, anyhow::Error>(NodeAddr::from_parts(
+                            id, server, vec![]
+                        ))
+                    })
+                    .collect::<std::result::Result<Vec<_>, _>>()
+                    .map_err(Into::into)
+                },
+            )
+            .await
+    }
+
+    /// Get the iroh secret key from the database or generate a new one and persist it.
+    pub(crate) async fn get_or_generate_iroh_keypair(&self) -> Result<SecretKey> {
+        match self.get_config_parsed(Config::IrohSecretKey).await? {
+            Some(key) => Ok(key),
+            None => {
+                let key = SecretKey::generate();
+                self.set_config_internal(Config::IrohSecretKey, Some(&key.to_string()))
+                    .await?;
+                Ok(key)
+            }
+        }
+    }
+
+    /// Get the iroh [NodeAddr] without direct IP addresses.
+    pub(crate) async fn get_iroh_node_addr(&self) -> Result<NodeAddr> {
+        let endpoint = self.iroh_endpoint.read().await;
+        let relay = endpoint
+            .as_ref()
+            .context("iroh endpoint not initialized")?
+            .my_relay();
+        Ok(NodeAddr::from_parts(
+            endpoint
+                .as_ref()
+                .context("iroh endpoint not initialized")?
+                .node_id(),
+            relay,
+            vec![],
+        ))
+    }
+
+    /// Get the topic for a given [MsgId].
+    pub(crate) async fn get_iroh_topic_for_msg(&self, msg_id: MsgId) -> Result<TopicId> {
+        let bytes: Vec<u8> = self
+            .sql
+            .query_get_value(
+                "SELECT topic FROM iroh_gossip_peers WHERE msg_id = ? LIMIT 1",
+                (msg_id,),
+            )
+            .await?
+            .context("couldn't restore topic from db")?;
+        Ok(TopicId::from_bytes(bytes.try_into().unwrap()))
+    }
+
+    /// Send realtime data to the gossip swarm.
+    pub async fn send_webxdc_realtime_data(&self, msg_id: MsgId, mut data: Vec<u8>) -> Result<()> {
+        let topic = self.get_iroh_topic_for_msg(msg_id).await?;
+        let has_joined = self.iroh_channels.read().await.get(&topic).copied();
+        if has_joined.is_none() {
+            self.send_webxdc_realtime_advertisement(msg_id).await?;
+        }
+
+        // depending on architecture this is 4 or 8 bytes.
+        // on some embedded deviced might be even more because of usize nature
+        let seq_num = has_joined.unwrap_or_default();
+        data.extend(seq_num.to_ne_bytes());
```

Comment:
why use native endian?

Reply:
no particular reason, this is just to duplicate anyway.

Reply:
then I would recommend using either big or little endian fixed, for cross machine compatability